docs/DesignDocs/system_ocall_opt_in.md

```diff
@@ -0,0 +1,120 @@
+if an enclave with edl file `sample.edl` wants to use the function `oe_log_ocall()` they can do
+one of the following:
+
+*Note: These examples assume that `${install_path}/include` is passed to edger8r as an edl search path*
+
+1. Import just the required ocall
+
+```
+// sample.edl
+enclave
+{
+    from "openenclave/edl/tee.edl" import oe_log_ocall;
+
+    trusted
+    {
+        public void my_sample_ecall();
+    };
+
+    untrusted
+    {
+        public void my_sample_ocall();
+    };
+}
+```
+
+2. Import all ocalls from `tee.edl`
+
+```
+// sample.edl
+enclave
+{
+    from "openenclave/edl/tee.edl" import *;
+
+    trusted
+    {
+        public void my_sample_ecall();
+    };
+
+    untrusted
+    {
+        public void my_sample_ocall();
+    };
+}
+```
+
+In example (1), linking would fail if the enclave made a call to `oe_realloc_ocall()`
+(also part of tee.edl) unless the developer modified the import line to
+`from "openenclave/edl/tee.edl" import oe_log_ocall, oe_realloc_ocall;`. In this scnario,
+example (2) will work without modification.
+
+## EDL local structures
+
+Under this design, system edl files will prefer using local structures; that is, structures
+that are defined in edl. This is distinct from foreign structures, which are defined in
+C/C++ header files and included in the edl file.
+
+In the case that a foreign structure must be used, system edl files may only include headers
+that are installed with the SDK.
+
+Local EDL structures provide additional security benefits detailed [here](
+https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/full_edl_serialization.md#motivation)
+so foreign structures should be avoided whenever possible.
+
+## Developer opt-out
+
+Because system ocalls are generally called by the SDK implementation, developers
+may not have direct control over which ocalls will be called by their application.
+A developer can refuse to import a function in EDL which will cause a linker error
+if it is used, but they should also be able to take action to resolve this failure.
+
+To this end, any OE SDK feature which makes an ocall on behalf of an enclave
+should provide a mechanism to disable the ocall at compile-time and fail
+gracefully.
+
+If this is not possible, the feature must be able to be disabled at the time
+of OE SDK compilation.
+
+## Exceptions to edl import and opt-in/out mechanisms
+
+A small number of ocalls in OE SDK are not implemented in edl and cannot be for
+technical reasons. These ocalls will always be included and the developer cannot
+opt out of them because they support core SDK functionality.
+
+These calls are:
+* `OE_OCALL_MALLOC`
+* `OE_OCALL_FREE`
+* `OE_OCALL_THREAD_WAIT`
+* `OE_OCALL_THREAD_WAKE`
+* `OE_OCALL_WAKE_HOST_WORKER`
+
+## List of system ocalls
+
+All system ocalls (other than the previously mentioned exceptions) are found
+in one of the following edl files
+
+* `common/tee.edl`
+* `common/syscall.edl`
+* `common/sgx/sgx.edl`
```

Comment:
This one is tricky because it's not really optional once you pull in oeenclave with its dependency on ocalls for sgx attestation quoting methods. `oe_get_cpuid_table_ocall()` is also always invoked on enclave creation, so you can't avoid including this.

Reply:
Technically attestation is optional though right? It might be a future work item but I think we could refactor to make attestation APIs opt in.

I'll look into the cpuid table, maybe this one actually is required. If it is the only one, we could move it to a separate EDL file and always include it (I don't want to force users to add an edl include line if they effectively have no choice).

Reply:
`oe_get_cpuid_table_ocall()` is always invoked today as part of oecore, not oeenclave. It shouldn't block this design, but as a matter of implementation there will be methods like this that should be removed. In this specific case, we can refactor the enclave initialization function to provide the required information in the initial ECall without having to rely on an OCall, which is also preferred from a performance perspective by reducing the number of boundary crossings.

If we do identify a subset of OE runtime OCalls that must always be included, we should probably submit a separate addendum to this design regarding how it those headers are handled. For example, one way of reducing the impact of such functions is to enforce that they are only invoked in codepaths that do not interact with the application's codepaths (e.g. only during enclave init or teardown).